Add Back button in file source navigation and ftp helper text

itisAliRH opened the pull request:
This PR adds the `Back` button for file source explorer when a navigation error and FTP helper text and oidc when `enable_oidc` in config is `true` when exploring FTP files.

Fixes the first and the second requirements in #12193

![Screen Shot 2021-12-25 at 7 55 10 PM](https://user-images.githubusercontent.com/8046843/147406871-6b810bcf-cca2-4ee3-9b57-6a59acb37352.png)
![Screen Shot 2021-12-25 at 6 52 30 PM](https://user-images.githubusercontent.com/8046843/147406883-740b6785-48ad-4686-880d-b593cf983529.png)
![Screen Shot 2021-12-25 at 6 48 21 PM](https://user-images.githubusercontent.com/8046843/147406894-5ab16e61-637a-46bc-8286-e4a4a0b38cb6.png)
![Screen Shot 2021-12-25 at 6 48 10 PM](https://user-images.githubusercontent.com/8046843/147406899-6c8c4209-eb0f-4728-ac04-465b4de98f45.png)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

itisAliRH committed:
Add Back button in for Chose Remote Data when there is an error navigation file source

itisAliRH committed:
Show ftp helper text in when navigating file source and oidc help text when enabled

itisAliRH committed:
Check and format code style

itisAliRH committed:
SelectionDialog: set default for `backFunc`, make `undoShow` optional and optional chaining for getGalaxyInstance to preventing raise an error during initializing

davelopez commented on client/src/components/FilesDialog/FilesDialog.vue:
```diff
@@ -258,10 +262,14 @@ export default {
             this.undoShow = !this.urlTracker.atRoot();
-            if (this.urlTracker.atRoot()) {
+            if (this.urlTracker.atRoot() || this.errorMessage) {
+                if (this.errorMessage) {
```
This if statement looks redundant since you don't want to show the errorMessage anyway you can directly set it to null.

itisAliRH replied:
That's right! I fixed it :)

davelopez commented on client/src/components/SelectionDialog/SelectionDialog.vue:
```diff
@@ -1,32 +1,44 @@
+                <strong>do not have a Galaxy password</strong> please use the reset password option in the login form
+                with your email to create a password for your account.</span
+            >
+        </b-alert>
```
This alert seems to be too specific to FTP. I think this `SelectionDialog` component is meant to be a bit more generic, so this information should be handled outside the component. Maybe use a slot to provide any kind of information message like this?

itisAliRH replied:
I agree with you. I created a new slot named `helper` for wrapping any help text for any resource. It would be nice to have a helper option in the `config/file_sources_conf.yml` config file.

davelopez replied:
Thank you!, I agree, the `showFtpHelper` could be even more generic to show other info messages from other file sources.

davelopez commented on client/src/components/SelectionDialog/SelectionDialog.vue:
```diff
@@ -1,32 +1,44 @@
                 <slot name="buttons"> </slot>
                 <b-btn size="sm" class="float-right selection-dialog-modal-cancel" @click="hideModal"> Cancel </b-btn>
             </div>
+            <div v-else class="w-100">
```
Not a big deal, but you probably can wrap this into a parent `div` so you don't have to duplicate the `class="w-100"`.

davelopez reviewed (commented):
Hi @itisAliRH!
Thank you for your contribution! Here are a couple of comments. Maybe other frontend colleagues can give you some more tips :)

itisAliRH committed:
Add new slot 'helper' to SelectionDialog and other improvements

in modal-fotter, wrap 'w-100' class to the parent div, delete showFtpHelper prop and data(), delete ftp alert

itisAliRH committed:
Wrap FTP alert to helper slot

Delete redundant if statement in the `load` method, add `ftpUploadSite` and `oidcEnabled` to data

davelopez commented on client/src/components/FilesDialog/FilesDialog.vue:
```diff
@@ -3,10 +3,24 @@
         :modal-show="modalShow"
-        :hide-modal="() => (modalShow = false)">
+        :hide-modal="() => (modalShow = false)"
+        :back-func="() => load()"
```
I believe it is considered better practice to emit an event from the child instead of passing callbacks to the child if possible :)

itisAliRH replied:
Yes, but I just wanted to follow code style (like `hide-modal`). So, I changed it.

itisAliRH committed:
FilesDialog: change `back-func` from callback. SelectionDialog: change back button @click

itisAliRH committed:
FilesDialog: add id `helper` for helper alert

itisAliRH committed:
testingData: add `ftpdId`, `empty-dir` in `rootResponse` and `someErrorText`

itisAliRH committed:
FilesDialog.test: Test show ftp helper and show loading error and return to root directory

davelopez reviewed (approved):
Nice!
Thank you for adding some tests too!
